collaboration/extending-kyma-cli.md

```diff
@@ -0,0 +1,346 @@
+
+Developer can modify this file to deploy or update a function on to the kyma cluster. This would enable the developer not only tune the function (by modifying replica, memory cpu), but also provide an overview of the current triggers for the function.
+
+Usually one does not change the config so often during development compared to code. So our aim should be to make deployment and testing the function simpler. So having a config file taking care of configuration would make deployment much simpler.
+
+Having such a config file would help developers with limited or no kubernetes knowledge to get themselves aquainted with various configuration options for function and concentrate on the business logic for writing the functions.
+
+Additionally it can be used automated deployments of functions. As this file can be read by kyma cli to deloy function with required configurations.
+
+Example structure of config.yaml
+
+```yaml
+name: <function_name>
+namespace: <namespace> #default by default
+events:
+  - event1
+  - event2
+environment-vars:
+  - foo:bar
+api:
+  url: <url>
+  actions:
+    - GET
+    - POST
+  security:
+    - JWT
+replicas: 1
+memory:
+  min:
+  max:
+cpu:
+  min:
+  max:
+runtime:
+debug-string:
+```
+
+##### Template files
+
+###### <function-code.extension>
+
+  Out of the box from the init command we should give an easy example which can be extended by the developer later. The example should be possible to be deployed directly on the kyma cluster. Here below there is a node example
+
+  ```js
+    module.exports = {
+      main: function (event, context) {
+        return 'Hello World';
+      }
+    };
+  ```
+
+###### <dependecies.extension>
+
+We should generate an empty package.json file which can be edited by developer as and when needed.
+
+###### index.js
+
+We should provide `index.js`(which would start the server and serve the handler function) out of the box which can be used by the developer to test his code locally.
+
+###### readme
+
+We should also provide a readme.md which would consist of following instructions:
+
+* How he can deploy the function into the cluster.
+* How he can run the function locally.
+* How to use the `config.yaml` file
+* Folder structure.
+* How to test debug it locally like using telepresence.
+
+##### Testing/Debugging
+
+*Local*
+
+We can introduce commands to test the function locally and also if the developer wants to debug the code. These can be achieved through make file.
+
+*Using Events from remote cluster*
+
+Sometimes the event payloads can be complex to be created locally. For such cases its nice to have the events on remote cluster triggering the code on the workstation. This would enable the developer to debug his code on the workstation using events being sent to the remote cluster. We can introduce command for example below:
+
+`kyma function debug <function-name> -n <namespace>`
```

Comment:
Allowing a distributed development environment is very important for application development (in our case writing a "function"). The sdk should support this, based on the eventing implementation which should be able to make a connection between the remote cluster and the local "eventing" part (the event generated on remote cluster should be propagated to the local "eventing"). We need to have something similar with kubectl port-forward but adapted to our context ( eventing-forward ...)

Reply:
we should discuss this in eventing section how can we achieve this local eventing communicating with remote cluster.